proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  with ``Stable2024`` (or even ``Experimental2024``) will be an error.
+  This policy encourages users to upgrade their editions in order
+  to access GHC's new features.
+
+- Once an edition has been eclipsed by newer models (that is, once the
+  calendar reads 2028 but someone is still using ``2024``), we know that
+  we are compiling older code. Our priority for such maintenance modes
+  is stability, not, say, making sure that the set of warnings conforms
+  to the latest standard.
+
+- We introduce several coarse-grained *semantic bundles* that group
+  together similar features. The individual choices are detailed below.
+  When using one of the new language
+  editions, error messages suggest opting into one of these bundles,
+  rather than suggesting individual extensions.
+
+  The bundles described below are all just combinations of existing
+  language extensions and warnings. I expect that bundles will
+  remain as such (though it's conceivable that we might imagine
+  language features that do not get their own extension, just a spot
+  in a bundle). It's also conceivable that the bundles will evolve
+  to encompass more expressive power (such as controlling optimization
+  flags or the meaning of ``import Prelude``).
+
+  Unlike language editions, a user may specify any number of bundles
+  when compiling a file. In order to simplify their processing, a
+  bundle cannot control the interpretation of other command-line
+  flags. Otherwise, though, a bundle can potentially control arbitrary
+  behavior of GHC, just like a language edition. The big difference
+  between them is that a file is compiled against exactly one language
+  edition, but it can be compiled against an arbitrary number of bundles.
+
+- When printing out the namne of a warning flag as part of a warning,
+  we also include any bundle that also controls the warning.
+
+Detailed specification
+~~~~~~~~~~~~~~~~~~~~~~
+
+The text above defines language editions generally. This section
+instantiates the general design with some specifics. These language
+editions affect language features (as described by extension name or
+warning flag in the chart below), as well as error messages (not described
+in any detail). They do not affect other aspects of GHC.
+
+The chart below classifies all current extensions and warnings.
+
+* The GHC2021 column has an X for extensions enabled as part of GHC2021.
+  For warnings, it lists if the warning is part of a current warning bucket.
+
+* The next four columns describe ``Student2024``, ``Stable2024``, ``Experimental2024``, and ``Latest``,
+  respectively. Here is the key for these columns:
+
+  * Y: an extension is turned on
+
+  * blank: the feature is available, but not on
+
+  * N: trying to enable the feature is an error
+
+  * W: the warning warns by default
+
+  * E: the warning is an error by default
+
+  * W!: the warning warns by default and cannot be turned off
```

Comment:
But this particular user has *asked* for this restriction, by saying `Stable2024`. If they want to disable the warning, just stop asking for the restriction, by saying `Experimental2024`.

Do you hate that Haskell rejects the following program?

```hs
stuff = [3, "a", 10, "b", 21, "c"]

odd acc1 acc2 [] = (acc1, acc2)
odd acc1 acc2 (x : xs) = even (acc1 + x) acc2 xs

even acc1 acc2 [] = (acc1, acc2)
even acc1 acc2 (x : xs) = odd acc1 (x : acc2) xs

(sum, letters) = odd 0 "" stuff
```

Executing this is "clearly possible": it results with `sum` being `34` and `letters` being `"cba"`, if I haven't made a mistake. But of course GHC will reject. And we *want* GHC to reject, because we have opted for a statically typed language. Similarly, a user opting for a stable language should be able to get what they want.

Reply:
I don't want to use `Experimental2024`. I want to use `Stable2024` with the warning turned off, which is a) pretty reasonable, since it clearly doesn't affect what the compiler can do) and b) something that I am used to from literally every other compiler I have ever used.

> Similarly, a user opting for a stable language should be able to get what they want.

What has _stability_ got to do with warning flags? Or being a student? Turning off a warning does not make my code unstable.

Reply:
If, for example, there was a warning whose behaviour was unstable, then perhaps we should not allow turning that warning on in `Stable` code... except again, we have disclaimed warning stability as a goal.